design/anova/anova.Rmd

```diff
@@ -0,0 +1,394 @@
+  mx <- component(fit, "x_matrix")
+  asg <- attr(mx, "assign")
+  formula <- fit$formula_parts$model_formula
+  tms <- terms(formula)
+  fcts <- attr(tms, "factors")[-1, , drop = FALSE] # discard the response
+  ods <- attr(tms, "order")
+  idx <- which(var == colnames(fcts))
+  cols <- which(asg == idx)
+  all_vars <- row.names(fcts)[which(fcts[, idx] == 1)]
+  all_levels <- vapply(all_vars, function(x) {
+    length(attr(fit$tmb_data$data[[x]], "levels"))
+  }, FUN.VALUE = 1)
+  total_levels <- prod(all_levels)
+  lvls <- prod(all_levels - 1)
+  l_mx <- matrix(0, nrow = lvls, ncol = ncol(mx))
+  l_mx[, cols] <- diag(rep(1, length(cols)))
+  for (i in seq_len(ncol(fcts))) {
+    x1 <- mx[, cols, drop = FALSE]
+    if (ods[i] >= ods[idx] && all(fcts[, i] >= fcts[, idx])) {
+      x0 <- mx[, which(!asg %in% c(i, idx)), drop = FALSE]
+      x2 <- mx[, which(asg == i), drop = FALSE]
+      m <- diag(rep(1, nrow(x0))) - x0 %*% solve(t(x0) %*% x0) %*% t(x0)
+      l_coefs <- solve(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x2
+      l_mx[, which(asg == i)] <- l_coefs
+    }
+  }
+  l_mx
+}
+
+get_l_matrix_type3 <- function(fit, var) {
+  mx <- component(fit, "x_matrix")
+  asg <- attr(mx, "assign")
+  formula <- fit$formula_parts$model_formula
+  tms <- terms(formula)
+  fcts <- attr(tms, "factors")[-1, , drop = FALSE] # discard the response
+  ods <- attr(tms, "order")
+  idx <- which(var == colnames(fcts))
+  cols <- which(asg == idx)
+  all_vars <- row.names(fcts)[which(fcts[, idx] == 1)]
+  all_levels <- vapply(all_vars, function(x) {
+    length(attr(fit$tmb_data$data[[x]], "levels"))
+  }, FUN.VALUE = 1)
+  total_levels <- prod(all_levels)
+  lvls <- prod(all_levels - 1)
+  l_mx <- matrix(0, nrow = lvls, ncol = ncol(mx))
+  l_mx[, cols] <- diag(rep(1, length(cols)))
+  for (i in seq_len(ncol(fcts))) {
+    if (ods[i] > ods[idx] && all(fcts[, i] >= fcts[, idx])) {
+      new_cols <- which(asg == i)
+      a_vars <- row.names(fcts)[which(fcts[, i] == 1)]
+      a_levels <- vapply(a_vars, function(x) {
+        length(attr(fit$tmb_data$data[[x]], "levels"))
+      }, FUN.VALUE = 1)
+      t_levels <- prod(a_levels)
+      l_mx[, which(asg == i)] <- l_mx[, which(asg == idx)] / t_levels * total_levels
+    }
+  }
+  l_mx
+}
+
+get_l_matrix_type2(fit, "ARMCD")
+get_l_matrix_type3(fit, "ARMCD")
+
+get_l_matrix_type2(fit, "AVISIT")
+get_l_matrix_type3(fit, "AVISIT")
+
+get_l_matrix_type2(fit, "ARMCD:AVISIT")
+get_l_matrix_type3(fit, "ARMCD:AVISIT")
+
+# call df_md on these l matrix to obtain the results.
+
+l_mt <- get_l_matrix_type3(fit, "AVISIT")
+df_md(fit, l_mt)
+```
+
+Note: design matrix involved with numerics are still not implemented yet.
+
+### Additional tries
+```{r}
+zap_small <- function(x) {
+  x[abs(x) < sqrt(.Machine$double.eps)] <- 0
+  x
+}
+
+fit <- mmrm(FEV1 ~ RACE + AVISIT + RACE * AVISIT + ar1(AVISIT | USUBJID), data = fev_data)
+
+x0 <- mx[, c(1, 4:6)]
+x1 <- mx[, 2:3, drop = FALSE]
+x2 <- mx[, c(7:12)]
```

Comment:
this fails, I guess column index needs to be corrected?

Reply:
I think I fixed it by adding line 380?

Reply:
oh yes. this just store some random code for testing, should be removed later (or just store it here for reference)